File: haskell/skylineProblem.hs
--       date: 2022-09-30
--      title: The Skyline Problem
--        url: https://leetcode.com/problems/the-skyline-problem/
-- difficulty: hard
import           Data.Array (Array, array, bounds, ixmap, listArray, (!))

data Point =
  Point Int Int
  deriving (Show, Eq)

data Building =
  Building Int Int Int
  deriving (Show)

getSkyline :: [Building] -> [Point]
getSkyline buildings = divideAndConquer buildingArray lowerBound upperBound
  where
    (lowerBound, upperBound) = (0, length buildings - 1)
    buildingArray = listArray (lowerBound, upperBound) buildings

-- Divide-and-conquer strategy which recursively splits buildings in half,
-- constructs skyline for each part and merges them.
-- Works with arrays for performance reasons.
divideAndConquer :: Array Int Building -> Int -> Int -> [Point]
divideAndConquer buildings start end
  | start > end = []
  | start == end = [Point xLeft height, Point xRight 0]
  | otherwise = pruneSkyline $ mergeSkylines left right
  where
    Building xLeft xRight height = buildings ! start
    mid = (start + end) `div` 2
    left = divideAndConquer buildings start mid
    right = divideAndConquer buildings (mid + 1) end

-- Removes consecutive points from a list with the same y-coordinate.
-- It is easier to first construct a skyline with redundant points,
-- and then apply pruneSkyline.
pruneSkyline :: [Point] -> [Point]
pruneSkyline [] = []
pruneSkyline (pt:pts) = pt : pruneSkyline' pts y
  where
    Point _ y = pt
    pruneSkyline' [] _ = []
    pruneSkyline' (Point x y:pts) previousY =
      if y == previousY
        then pruneSkyline' pts previousY
        else Point x y : pruneSkyline' pts y

-- Merges two skylines constructed for different buildings.
mergeSkylines :: [Point] -> [Point] -> [Point]
mergeSkylines firstList secondList = mergeSkylines' firstList 0 secondList 0

mergeSkylines' :: [Point] -> Int -> [Point] -> Int -> [Point]
mergeSkylines' [] _ [] _ = []
mergeSkylines' [] firstH secondList _ =
  map (\(Point x y) -> Point x $ max firstH y) secondList
mergeSkylines' firstList _ [] secondH =
  map (\(Point x y) -> Point x $ max secondH y) firstList
mergeSkylines' firstList firstH secondList secondH =
  Point (min firstX secondX) (max firstH' secondH') :
  mergeSkylines' firstList' firstH' secondList' secondH'
  where
    Point firstX firstY:firstTail = firstList
    Point secondX secondY:secondTail = secondList
    (firstList', firstH', secondList', secondH')
      | firstX > secondX = (firstList, firstH, secondTail, secondY)
      | firstX < secondX = (firstTail, firstY, secondList, secondH)
      | firstX == secondX = (firstTail, firstY, secondTail, secondY)

{-
import Control.Exception.Base (assert)

buildings1 = [Building 2 9 10, Building 3 7 15, Building 5 12 12, Building 15 20 10, Building 19 24 8]
expectedSkyline1 = [Point 2 10, Point 3 15, Point 7 12, Point 12 0, Point 15 10, Point 20 8, Point 24 0]
assert (expectedSkyline1 == getSkyline buildings1) ()

buildings2 = [Building 0 2 3, Building 2 5 3]
expectedSkyline2 = [Point 0 3, Point 5 0]
assert (expectedSkyline2 == getSkyline buildings2) ()
-}

File: src/main/kotlin/skylineProblem.kt
//       date: 2022-09-30
//      title: The Skyline Problem
//        url: https://leetcode.com/problems/the-skyline-problem/
// difficulty: hard
//------------------------------------------------------------------------------

fun getSkyline(buildings: Array<IntArray>): List<IntArray> =
    mapToList(Point::toIntArray, divideAndConquer(buildings.map(::intArrayToBuilding).toTypedArray()))

//------------------------------------------------------------------------------
// A couple of classes to make data more readable

private class Building(val leftX: Int, val rightX: Int, val height: Int)

private fun intArrayToBuilding(arr: IntArray): Building =
    Building(arr[0], arr[1], arr[2])

private class Point(val x: Int, val y: Int) {
    fun toIntArray(): IntArray = intArrayOf(x, y)
    operator fun component1(): Int = x
    operator fun component2(): Int = y
}

//------------------------------------------------------------------------------
// Let's reimplement Lisp lists :-)
// The code with Kotlin's List<T> runs out of time for me.

private class Cons<T>(val car: T, val cdr: Cons<T>? = null) {
    operator fun component1(): T = car
    operator fun component2(): Cons<T>? = cdr
}

private fun <T, R> map(f: (T) -> R, list: Cons<T>?): Cons<R>? =
    if (list == null) null else Cons(f(list.car), map(f, list.cdr))

private fun <T, R> mapToSequence(f: (T) -> R, list: Cons<T>?): Sequence<R> =
    sequence {
        if (list != null) {
            yield(f(list.car))
            yieldAll(mapToSequence(f, list.cdr))
        }
    }

private fun <T, R> mapToList(f: (T) -> R, list: Cons<T>?): List<R> =
    mapToSequence(f, list).toList()

//------------------------------------------------------------------------------

// Divide-and-conquer strategy which recursively splits buildings in half,
// constructs skyline for each part and merges them.
// Works with arrays for performance reasons.
private fun divideAndConquer(
    buildings: Array<Building>,
    start: Int = 0,
    end: Int = buildings.lastIndex
): Cons<Point>? =
    if (start > end) {
        null
    } else if (start == end) {
        val firstPair = Point(buildings[start].leftX, buildings[start].height)
        val secondPair = Point(buildings[start].rightX, 0)
        Cons(firstPair, Cons(secondPair, null))
    } else {
        val mid = (start + end) / 2
        val left = divideAndConquer(buildings, start, mid)
        val right = divideAndConquer(buildings, mid + 1, end)
        pruneSkyline(mergeSkylines(left, right))
    }

// Removes consecutive points from a list with the same y-coordinate.
// It is easier to first construct a skyline with redundant points,
// and then apply pruneSkyline.
private fun pruneSkyline(list: Cons<Point>?): Cons<Point>? =
    if (list == null) null else Cons(list.car, pruneSkylineRec(list.cdr, list.car.y))

private fun pruneSkylineRec(list: Cons<Point>?, previousY: Int): Cons<Point>? =
    when {
        list == null -> null
        list.car.y == previousY -> pruneSkylineRec(list.cdr, previousY)
        else -> Cons(list.car, pruneSkylineRec(list.cdr, list.car.y))
    }

// Merges two skylines constructed for different buildings.
private fun mergeSkylines(
    firstList: Cons<Point>?,
    secondList: Cons<Point>?,
    firstH: Int = 0,
    secondH: Int = 0
): Cons<Point>? =
    when {
        firstList == null && secondList == null ->
            null

        firstList == null ->
            map({ point: Point -> Point(point.x, Math.max(firstH, point.y)) }, secondList)

        secondList == null ->
            map({ point: Point -> Point(point.x, Math.max(secondH, point.y)) }, firstList)

        else -> {
            val (firstHead, firstTail) = firstList
            val (firstX, firstY) = firstHead
            val (secondHead, secondTail) = secondList
            val (secondX, secondY) = secondHead

            val newFirstList: Cons<Point>?
            val newFirstH: Int
            val newSecondList: Cons<Point>?
            val newSecondH: Int

            when {
                firstX > secondX -> {
                    newFirstList = firstList
                    newFirstH = firstH
                    newSecondList = secondTail
                    newSecondH = secondY
                }

                firstX < secondX -> {
                    newFirstList = firstTail
                    newFirstH = firstY
                    newSecondList = secondList
                    newSecondH = secondH
                }

                else -> {
                    newFirstList = firstTail
                    newFirstH = firstY
                    newSecondList = secondTail
                    newSecondH = secondY
                }
            }

            val skylineX = Math.min(firstX, secondX)
            val skylineY = Math.max(newFirstH, newSecondH)

            Cons(Point(skylineX, skylineY), mergeSkylines(newFirstList, newSecondList, newFirstH, newSecondH))
        }
    }
